Explore: Add functionality to show/hide query row results

ivanahuckova opened the pull request:
**What this PR does / why we need it**:
This PR introduces new functionality to disable/enable query (show/hide query row results). __Clear row__ button has been replaced with button to toggle disable/enable query. `changeQuery` function has been used to target single query and change its hide property to true/false. Also, displayed latency in query row is set to 0 if query results are hidden. This makes it more user friendly - as users are not confused with latency in hidden queries. Also, this makes live tailing more straight forward - users know that hidden queries are not affecting live logs. 

![ezgif com-video-to-gif (12)](https://user-images.githubusercontent.com/30407135/68545948-442c7080-03d2-11ea-954f-130fa055f826.gif)

**Which issue(s) this PR fixes**:
Fixes #18781

ivanahuckova committed:
Explore: Hide queries - all or nothing visibility

ivanahuckova committed:
Explore: toggle hide/show query

ivanahuckova committed:
Explore: Refactor code to simplify the functionality and remove unnecessary state

ivanahuckova committed:
Explore: Set latency to 0 if query is hidden

aocenas commented on public/app/features/explore/QueryRow.tsx:
```diff
@@ -76,8 +76,10 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
-    this.onChange(null, true);
+  onClickToggleHiddenQuery = (query: DataQuery) => {
+    const { exploreId, index } = this.props;
+    !!this.props.query.hide ? (query.hide = false) : (query.hide = true);
```
We are mutating props give to us by parent here which can create lots of problems. Better just create a new query and pass that up:
```
const newQuery = {
  ...this.props.query
  hide: !this.props.query.hide
};
this.props.changeQuery(exploreId, newQuery, index, true);
```

This also means you do not have to pass `query` arg here as it always works on `this.props.query`. Also in general it is confusing when we use in the same context both `query` arg and `this.props.query` which both point to the same object.

ivanahuckova replied:
Yes. Makes so much sense. Thanks Andrej. It is fixed now. 🙂

aocenas commented on public/app/features/explore/QueryRow.tsx:
```diff
@@ -174,8 +176,11 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
+              className="gf-form-label gf-form-label--btn"
+              onClick={() => this.onClickToggleHiddenQuery(this.props.query)}
+            >
+              <i className={`${!!this.props.query.hide ? 'fa fa-eye-slash' : 'fa fa-eye'}`} />
```
No need to wrap in back-tick quotes here

ivanahuckova replied:
🤦‍♀️yes, not sure why I did this. Fixed as well.

ivanahuckova committed:
Explore: Use newQuery instead of mutating prop

ivanahuckova committed:
Explore: Refactor

davkal reviewed (changes requested):
Works really well on fast queries in Prometheus. Found some issues though:

Blockers: 

- users want to show/hide quickly, triggering a new query that takes long will not work from a UX point of view (hide existing results instead)
- weird behavior in loki: run a query, then hide all the results -> loading indicator keeps showing and results never go away
- when switching datasources that support imports (eg. Prometheus -> Loki), the hide state of the row is maintained

Minor: 

- hiding an empty metrics query removes the cheat sheet

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivanah/explore/hide-query

ivanahuckova committed:
Add hiddenQuery state to separate visual hiddenQuery representation and query execution

ivanahuckova commented:
> * users want to show/hide quickly, triggering a new query that takes long will not work from a UX point of view (hide existing results instead)

Yes, I totally agree David. However this is problematic, as series that are returned from database don't have refIds. Therefore it is not easy to hide them, based on the queryRow's refId. This will be even more problematic for logs. 

Moreover, the same implementation is currently done in dashboards. So if we would like to implement it differently, it would probably be a bigger project and we would probably need to do the same for the dashboards. 

So if we decide to do it, we should probably consult this with dashboard squad. 

> * weird behavior in loki: run a query, then hide all the results -> loading indicator keeps showing and results never go away

This was fixed here: https://github.com/grafana/grafana/pull/19830

> * when switching datasources that support imports (eg. Prometheus -> Loki), the hide state of the row is maintained
Fixed 🙂

> * hiding an empty metrics query removes the cheat sheet
Fixed 🙂

torkelo commented:
> Yes, I totally agree David. However this is problematic, as series that are returned from database don't have refIds

There can be refIds on each seres (DataFrame) to know which query it came from. The problem is t that this concept is quite new (~1-2 years) and very few data sources actually set refId on the returned queries. And for some data sources like Graphite it is currently impossible to know which query a series came from as all queries are issued in one request and the return structure does not let grafana figure out which series is the result of which query. There could be a similar problem for other data sources as well.

davkal commented:
We had this previously abstracted in Explore where a click on Run would create a series of requests (called transactions) that each kept references to their results (see code comment from v6.0.0):
https://github.com/grafana/grafana/blob/34a9a621b6b6f14d9ad15690d0a38047905b5423/public/app/features/explore/Explore.tsx#L74
Then a click on show/hide would have re-run the result merging, but not trigger a query. That way we did not have to rely on datasource specific handling of results. To the datasource it was a single operation. Explore was in charge of managing this.

davkal commented:
> Moreover, the same implementation is currently done in dashboards.

That's true. I was just hoping to be even more responsive. But given the consistent behavior. I no longer consider this a blocker.

davkal commented:
Weird observation in this branch: when reloading the tab with Explore and a loki query present, the query does not execute. But it seems to work in master and there is nothing in the PR that seems to  be related.

ivanahuckova committed:
Merge branch 'master' into ivanah/explore/hide-query

ivanahuckova committed:
Explore: Remove duplicated state, fix starting issue

ivanahuckova commented:
@davkal @aocenas 
I've found an issue - I was initialising hiddenQuery state with false value, but if you would send a link to someone else/refresh page, query.hide could be true, and it would make hiddenQuery state and query.hide unsynced. What I decided to do is to remove hiddenQuery state and use just query.hide. Also, I've added *if* statement in the function -> if you have initial state and query input is empty, you can not toggle hide/show query response, as there are no data coming in and instead cheatsheets are displayed.

davkal commented:
Behavior looks good to me now.

ivanahuckova committed:
Merge remote-tracking branch 'origin' into ivanah/explore/hide-query

ivanahuckova committed:
Explore: Refactor, add tests

aocenas commented on public/app/features/explore/QueryRow.tsx:
```diff
@@ -76,8 +77,15 @@ export class QueryRow extends PureComponent<QueryRowProps, QueryRowState> {
-    this.onChange(null, true);
+  onClickToggleHiddenQuery = () => {
+    const { exploreId, index, query, queryResponse } = this.props;
+    if (queryResponse.state !== LoadingState.NotStarted) {
```
What about moving this inside the queryActions so that the show/hide button is disabled in this case?

aocenas commented on public/app/features/explore/QueryRowActions.test.tsx:
```diff
@@ -0,0 +1,38 @@
+  });
+  it('should render component without editor mode', () => {
+    const wrapper = setup({ canToggleEditorModes: false });
+    expect(wrapper.find('i.fa-pencil')).toHaveLength(0);
```
One way to also do this to check for elements is to use `aria-labels`. Those should be more stable (ie `query edit` label should be more stable than particular icon class). Not sure if it is applicable later in the show/hide button tests if we want to test if icon actually changes but in general when finding some particular element in the DOM, `aria-label` is preferred to classes or IDs, and also has the benefit of better accessibility.

ivanahuckova committed:
Refactor, disable hide query button when cheatsheet displayed
